active/0015-unified-authentication.md

```diff
@@ -0,0 +1,174 @@
+1. If there are more than one auth plugin enabled, there is no deterministic order for
+   how the different backends are checked.
+1. Enabled authentication plugins are collectively considered one global instance,
+   there is a lack of granularity for scoped control levels. e.g. per-zone, or per-listener.
+
+To address the pain-points in 5.0, we propose below enhancements.
+
+## Design
+
+### One app for all
+
+One `emqx_authn` app to unify the management of all different backends (except for ldap being postponed for now).
+
+### The same hook-point
+
+In this design, there is no intention to change how EMQ X hooks work,
+the new app `emqx_authn` will continue to make use of the `client.authenticate` hook-point,
+only to dispatch auth requests to the underlying backends inside one single hook call.
+
+### Composable authn "chain"
+
+We should allow users to compose (configure) a "chain" of backends with a defined order in which the
+checks are performed one after another. Each check against the backend in the chain may yield 3 different
+results for one-request authentication:
+
+- `ignore` is to indicate there is no auth information found hence should
+   continue validate the client against the rest of the backends in the chain.
+- `{ok, Info}` as a login accepted, hence to terminate the auth calls from here,
+   where `Info` may contain additional information such as to indicate if the user is a super-user.
+- `{error, Reason}` to indicate that client's login should be denied.
+
+NOTE: for temporary errors, such as database connection issue, the error is logged,
+      and the auth result is `ignore` so to move forward to the next node in the chain.
+
+NOTE: if there is no `ok` (accepted) result after a full chain exhaustion, the login is rejected.
+
+For enhanced authentication, such as `scram` there can be messages after the first request,
+hence the backend may return `{continue, Data}`,
+where `Data` is to be kept by the connection process as handling context for the following messages.
+
+### Fine-grained configuration levels
+
+By default, EMQ X user can configure one global chain which applies to all MQTT listeners,
+we should however also allow a per-listener configuration to override the global chain.
+Together with firewall rules, this will allow users to have different auth solution for
+MQTT service facing different group of clients coming from their designated network.
+
+### Reconfigurable on the fly
+
+The changes in the auth chain or the backends should be applied on-the-fly
+i.e. without having to restart the `eqmx_authn` application.
+
+
+## Configuration
+
+- Example config for built-in-database (mnesia) username/password based global auth
+
+```
+authentication {
+  backend: 'built-in-database',
+  mechanism: "password-based",
+  ...
+  user_id_type: clientid
+}
+```
+
+- Example config for built-in-database (mnesia) username/password based per-listener auth
+
+```
+listener.tcp.default {
+  ...
+  authentication: {
```

Comment:
In this example and the example above we have single authenticator config as `authentication` value; but due to the proposal we may have a chain of authenticators, shouldn't we have a list here?